### PR TITLE
Dpos epoch id delegate trie

### PR DESCRIPTION
**Pull request by xfliudx**
## Description

* Add epoch ID into the delegate trie as the prefix when it is updated, get, delete.
* Genesis block delegate  trie doesn't exist the prefix due to its epoch ID changing with user's setting in genesis.json.  

## Type of change

- [X] Bug Fix
- [X] New Feature

## Testing

Fix the incompatible bugs

### CODE REVIEW

File: consensus/dpos/dpos.go [hunk 1]
```diff
@@ -388,18 +389,31 @@ func accumulateRewards(config *params.ChainConfig, state *state.StateDB, header
...
+	epochID := CalculateEpochID(header.Time.Int64())
+	epochBytes := common.Int64ToBytes(epochID)
+
 	delegateTrie := dposContext.DelegateTrie()
-	delegatorIterator := trie.NewIterator(delegateTrie.PrefixIterator(header.Validator.Bytes()))
-	for delegatorIterator.Next() {
+	delegatorIterator := trie.NewIterator(delegateTrie.PrefixIterator(append(epochBytes, header.Validator.Bytes()...)))
+	isExist := delegatorIterator.Next()
+
+	if !isExist {
```
[the remark this answers is not in the document]
Comment (jwangdx) [reply consensus/dpos/dpos.go]: You could further refactor the code here to the following

```
const genesisEpochID int64 = 0

var errNoEntriesInDelegateTrie = errors.New("no entries in delegate trie")

func forEachDelegateForCandidate(ctx *types.DposContext, epochID int64, candidate common.Address, cb func(delegator common.Address) {
	delegateTrie := ctx.DelegateTrie()
	delegatorIterator := trie.NewIterator(delegateTrie.PrefixIterator(makeDelegateTriePrefix(epochID, candidate)))
	var hasEntry bool
	for delegatorIterator.Next() {
		hasEntry = true
		delegator := common.BytesToAddress(delegatorIterator.Value)
		if err := cb(delegator); err != nil {
			return err
		}
	}
	if !hasEntry {
		return errNoEntryInDelegateTrie
	}
	return nil
}

func accumulateRewards(config *params.ChainConfig, state *state.StateDB, header *types.Header, dposContext *types.DposContext) error {
	...
	epochID := CalculateEpochID(header.Time.Int64())
	assignedReward := common.BigInt0
	// define the function to assign reward to delegator
	assignReward := func(delegator common.Address) error {
		delegatorVote := getVoteLastEpoch(state, delegator)
		delegatorReward := delegatorVote.Mult(sharedReward).Div(voteCount)
		state.AddBalance(delegator, delegatorReward.BigIntPtr())
		assignedReward = assignedReward.Add(delegatorReward)
	}
	err := forEachDelegateForCandidate(ctx, epochID, header.Validator, assignReward)
	if err == errNoEntryInDelegateTrie {
		epochID = genesisEpochID
		err = forEachDelegateForCandidate(ctx, epochID, header.Validator, assignReward)
	}
	if err != nil {
		return err
	}
	validatorReward := blockReward.Sub(assignedReward)
	state.AddBalance(header.Coinbase, validatorReward.BigIntPtr())
	return nil
}
``` 

The advantage of this refactoring is that the business logic and the read/write method is defined in two different functions, which gives much more extensibility to the code. 

Furthermore, the `forEachDelegateForCandidate` method could be moved to a different module (e.g. types/dpos_context.go) thus could modularize the `DposContext`. If one want to do some operation on the delegateTrie, he does not need to know the exact implementation of delegate trie. Instead, to use the function `forEachDelegateForCandidate` is the only thing he needs to know.

File: consensus/dpos/epoch_context_test.go [hunk 2]
```diff
@@ -173,7 +175,7 @@ func Test_KickoutValidators(t *testing.T) {
 			t.Errorf("failed to delete the kick out one from candidate trie: %s", candidates[i].String())
 		}
 
-		delegatorFromTrie := epochContext.DposContext.DelegateTrie().Get(append(candidates[i].Bytes(), delegator.Bytes()...))
+		delegatorFromTrie := epochContext.DposContext.DelegateTrie().Get(append(epochIdBytes, append(candidates[i].Bytes(), delegator.Bytes()...)...))
```
Comment (jwangdx) [inline consensus/dpos/epoch_context_test.go]: Please avoid defining the key multiple times

File: core/types/dpos_context.go [hunk 3]
```diff
@@ -224,10 +224,12 @@ func (dc *DposContext) KickoutCandidate(candidateAddr common.Address) error {
 		}
 	}
 
-	iter := trie.NewIterator(dc.delegateTrie.PrefixIterator(candidate))
+	epochBytes := common.Int64ToBytes(epochID)
+
+	iter := trie.NewIterator(dc.delegateTrie.PrefixIterator(append(epochBytes, candidate...)))
```
Comment (jwangdx) [inline core/types/dpos_context.go]: Please avoid defining the prefix every time used.

File: core/types/dpos_context.go [hunk 4]
```diff
@@ -224,10 +224,12 @@ func (dc *DposContext) KickoutCandidate(candidateAddr common.Address) error {
...
 	for iter.Next() {
 		delegator := iter.Value
-		key := append(candidate, delegator...)
+		key := append(epochBytes, append(candidate, delegator...)...)
```
Comment (jwangdx) [inline core/types/dpos_context.go]: You can obtain the key by `iter.Key`. No need to define it once more.

File: core/types/dpos_context.go [hunk 5]
```diff
@@ -316,7 +319,7 @@ func (dc *DposContext) Vote(delegatorAddr common.Address, candidateList []common
 		}
 
 		for _, oldCandidate := range oldCandidateList {
-			err = dc.delegateTrie.TryDelete(append(oldCandidate.Bytes(), delegator...))
+			err = dc.delegateTrie.TryDelete(append(epochBytes, append(oldCandidate.Bytes(), delegator...)...))
```
Comment (jwangdx) [inline core/types/dpos_context.go]: Do not define again.

File: core/types/dpos_context.go [hunk 6]
```diff
@@ -344,7 +347,7 @@ func (dc *DposContext) Vote(delegatorAddr common.Address, candidateList []common
 			continue
 		}
 
-		err = dc.delegateTrie.TryUpdate(append(candidate, delegator...), delegator)
+		err = dc.delegateTrie.TryUpdate(append(epochBytes, append(candidate, delegator...)...), delegator)
```
Comment (jwangdx) [inline core/types/dpos_context.go]: Do not define again.

File: core/types/dpos_context.go [hunk 7]
```diff
@@ -394,7 +398,7 @@ func (dc *DposContext) CancelVote(delegatorAddr common.Address) error {
 
 	// delete all vote records from delegateTrie
 	for _, oldCandidate := range oldCandidateList {
-		err = dc.delegateTrie.TryDelete(append(oldCandidate.Bytes(), delegator...))
+		err = dc.delegateTrie.TryDelete(append(epochBytes, append(oldCandidate.Bytes(), delegator...)...))
```
Comment (jwangdx) [inline core/types/dpos_context.go]: Do not define again.

File: core/types/dpos_context_test.go [hunk 8]
```diff
@@ -90,7 +93,7 @@ func TestDposContextKickoutCandidate(t *testing.T) {
 	}
 
 	for i, candidate := range candidates {
-		delegateIter := trie.NewIterator(dposContext.delegateTrie.PrefixIterator(candidate.Bytes()))
+		delegateIter := trie.NewIterator(dposContext.delegateTrie.PrefixIterator(append(epochIdBytes,candidate.Bytes()...)))
```
Comment (jwangdx) [inline core/types/dpos_context_test.go]: Do not define again.

File: core/types/dpos_context_test.go [hunk 9]
```diff
@@ -120,15 +123,15 @@ func TestDposContextVoteAndCancelVote(t *testing.T) {
...
-	_, err = dposContext.Vote(delegator, []common.Address{notExistAddress})
+	_, err = dposContext.Vote(delegator, []common.Address{notExistAddress}, epochID)
 	assert.NotNil(t, err)
 
 	// delegator vote to old candidate
-	_, err = dposContext.Vote(delegator, []common.Address{candidate})
+	_, err = dposContext.Vote(delegator, []common.Address{candidate}, epochID)
 	assert.Nil(t, err)
-	delegateIter := trie.NewIterator(dposContext.delegateTrie.PrefixIterator(candidate.Bytes()))
+	delegateIter := trie.NewIterator(dposContext.delegateTrie.PrefixIterator(append(epochIdBytes,candidate.Bytes()...)))
```
Comment (jwangdx) [inline core/types/dpos_context_test.go]: Do not define again.

File: consensus/dpos/dpos.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwangdx) [reply consensus/dpos/dpos.go]: The first block of an epoch (the block where tryElect is executed) is produced by a validator in the last epoch. Thus the delegate data should use the delegate in the last epoch, right? In this special case, the epoch should be calculated using `CalculateEpochID(curTime)-1`.

Furthermore, considering the first block in an epoch, should `accumulateRewards` be executed before `tryElect` in function `Finalize`?

**Comment by jwangdx** [comment]
Please also fix the formatting issues in the Codacy report.

File: consensus/dpos/dpos.go [hunk 10]
```diff
@@ -388,18 +389,31 @@ func accumulateRewards(config *params.ChainConfig, state *state.StateDB, header
...
 	assignedReward := common.BigInt0
+
 	// Loop over the delegators to add delegator rewards
+	epochID := CalculateEpochID(header.Time.Int64())
+	epochBytes := common.Int64ToBytes(epochID)
+
 	delegateTrie := dposContext.DelegateTrie()
-	delegatorIterator := trie.NewIterator(delegateTrie.PrefixIterator(header.Validator.Bytes()))
-	for delegatorIterator.Next() {
+	delegatorIterator := trie.NewIterator(delegateTrie.PrefixIterator(append(epochBytes, header.Validator.Bytes()...)))
```
[the remark this answers is not in the document]
Comment (xfliudx) [reply consensus/dpos/dpos.go]: Cool, agree.

**Pull request closed without merging**